apkenv.c

```diff
@@ -227,6 +243,82 @@ install_overrides(struct SupportModule *module)
     }
 }
 
+void
+dalvikvm_JNIEnv_DeleteLocalRef(JNIEnv* p0, jobject p1)
+{
+#ifdef APKENV_DEBUG
+    printf("dalvikvm_JNIEnv_DeleteLocalRef(%x)\n", p1);
+#endif
+    global.dalvik_copy_env.DeleteLocalRef(p0, p1);
+    (*p0)->ExceptionClear(p0);
+}
+
+
+static void
+install_default_dalvik_overrides()
+{
+    struct JNINativeInterface *dalvik_env = (struct JNINativeInterface *)global._env->functions;
+
+    int pagesize = sysconf(_SC_PAGESIZE);
+    mprotect((char*)((int)dalvik_env & ~(pagesize-1)), pagesize*2, PROT_READ | PROT_WRITE);
```

Comment:
Might be a bit more readable if done like this (untested, the modulo math might not be correct):

```
char *dalvik_env_page_aligned = (char *)dalvik_env - ((char *)dalvik_env % pagesize);
mprotect(dalvik_env_page_aligned, pagesize*2, PROT_READ | PROT_WRITE);
// ...
mprotect(dalvik_env_page_aligned, pagesize*2, PROT_READ);
```

Also, this assumes `sizeof(struct JNINativeInterface) < pagesize`. From reading the [mprotect(2)](http://man7.org/linux/man-pages/man2/mprotect.2.html) manpage, it looks like while the `addr` parameter needs to be page-aligned, `len` doesn't need to be a multiple of the page, so `mprotect(dalvik_env_page_aligned, sizeof(*dalvik_env) + ((char *)dalvik_env % pagesize), PROT_READ | PROT_WRITE);` could be enough (again, untested -- just thinking aloud here).